### ISSUE TITLE
Progress bar with '...'

### ISSUE THREAD

**Issue by brupelo** [tqdm/tqdm#925]
Hi, I'd love to start using this library, is there any easy way to print the typical '...' progress bar? Ie, something like this:

    import time
    import threading
    import sys

    class ProgressBar():

        def __init__(self):
            self.start = time.time()
            self.count = 0
            self.completed = False
            self.width = 4

        def stop(self):
            self.completed = True

        def print(self, msg):
            while not self.completed:
                elapsed = time.time() - self.start
                if elapsed > 0.3:
                    self.start = time.time()
                    self.count += 1
                    suffix = '.' * (self.count % self.width)
                    suffix = suffix.ljust(self.width - self.count % self.width, ' ')
                    print("{}{}".format(msg, suffix), end='\r', flush=True)

    if __name__ == '__main__':
        pb = ProgressBar()
        t = threading.Thread(target=pb.print, args=("Doing stuff",))
        t.start()
        time.sleep(5)
        pb.stop()
        t.join()

**Comment by casperdcl** [comment]
you mean some variant of a spinner? Except instead you havr dots filling up, resetting and refilling constantly?

**Comment by brupelo** [comment]
@casperdcl Yeah, that's right, a spinner... at first I'd considered this library but then someone from #python just pointed me out for this particular task `yaspin` was a better choice.. 

I guess a spinner is not the main use-case of `tqdm`...? or is it? Please let me know or just close if you think this issue doesn't apply here.

Thanks!

**Comment by casperdcl** [comment]
It's not a use-case of `tqdm` but

1. It's been reqeiested several times
2. I can't find a good open issue

So will probably leave this open.

**Comment by nishnash54** [comment]
We can have a simple spinner after the bar, something like [alive progress](https://github.com/rsalmei/alive-progress). That way we can optionally enable/disable the bar/spinner. It would be helpful in-terms of showing the process is alive when running computationally expensive tasks with long runtimes.

**Comment by casperdcl** [comment]
Think we need to decide on a user interface first.


```python
from tqdm.auto import tqdm

# format for `ascii`: empty_char + spinner_chars + error_char + success_char
with tqdm.spinner(desc="Processing", ascii=" |/-\\!✔") as t:
    try:
        some_long_running_process()
    except:
        print("the spinner should catch this error an display `!`")
        raise
```

Pros:

- Easy to implement and use

Cons:

- Doesn't allow combining progress bar with spinner (e.g. operations with a long time between `update()`s

----

```python
with tqdm(desc="Processing", ascii=" |/-\\!✔") as t:
    t.spin()  # starts a thread
    try:
        some_long_running_process()
    except:
        print("the spinner should catch this error an display `!`")
        raise
```

Pros:

- allows room for more customisation (e.g. bar & spinner): `t.spin()` could have lots of arguments without polluting `tqdm.__init__` further.

Cons:

- more annoying to implement

**Comment by lrq3000** [comment]
Just for info, there was a quite old PR implementing both single-line and multi-line spinners: #223 

If I remember correctly, the goal there was a bit more ambitious, to allow the user to implement/reproduce any kind of spinner or custom progress bar as can be seen in various other libraries.

But maybe you'd like something simpler to implement, as I don't know how it would fit in the current tqdm architecture. Also maybe the interface is not the easiest, that would surely need a refresh now that tqdm is mature.

**Comment by lrq3000** [comment]
Another bit of info about this PR: the single-line progress bars / spinners were fully implemented, only the multi-line variant still needed some fixing (it was working but there were some edge cases notably when the progress bar finishes that were not managed correctly, eg after leave=True).

**Comment by casperdcl** [comment]
Completely forgot about #223; looks like something intended for `tqdm.contrib` and if it gets used a lot made part of `tqdm.std`.

Was thinking about something wayyy simpler though with https://github.com/tqdm/tqdm/issues/925#issuecomment-623142879

**Comment by GlenNicholls** [comment]
>We can have a simple spinner after the bar, something like alive progress. That way we can optionally enable/disable the bar/spinner. It would be helpful in-terms of showing the process is alive when running computationally expensive tasks with long runtimes.

I strongly agree with the above! I also have another use-case that I realized isn't currently possible with tqdm (at least to my knowledge):

At my company, we build our FPGA designs using Python `subprocess` to run the desired tool from the CLI. Our builds can take 12+ hours. Right now, we print all the build output to stdout to show that progress is being made, but it masks sometimes important python logs. I asked https://stackoverflow.com/questions/70028231/how-to-set-up-progress-bar-and-also-print-output-from-subprocess-popen because I would like my colleagues to see that progress is being made, have an idea of run-time, and also see the terminal output on a single line so they know approximately where the build is at. With these things combined, it wouldn't clutter the terminal, we would know approximately how much longer until a build finishes, and I wouldn't need to hackishly rely on terminal output to tell tqdm to increment the progress counter...

**Comment by pbk0** [comment]
Any progress on this ... I am not fond of having two dependencies `tqdm` and `yapin` ... also `alive-progress` sounds cool but not fond of refactoring now ... also don't know the pitfalls there given that `tqdm` is mature solution

**Comment by GlenNicholls** [comment]
@pbk0 I switched to rich. Single dependency and it handles everything terminal related.

**Comment by pbk0** [comment]
> @pbk0 I switched to rich. Single dependency and it handles everything terminal related.

That looks impressive ... will look at that

**Comment by cblackbu** [comment]
> # format for `ascii`: empty_char + spinner_chars + error_char + success_char
> with tqdm.spinner(desc="Processing", ascii=" |/-\\!✔") as t:
>     try:
>         some_long_running_process()
>     except:
>         print("the spinner should catch this error an display `!`")
>         raise

I would like to see something that only animates after proof that the thread hasn't actually frozen.  That is an option where the spinner only spins on t.update().  Ideally long running processes are repetitive processes and could call update() every iteration or every 100th iteration, whatever makes sense.
